### PR TITLE
osd: report slow requests and pending creating pgs to mgr

### PR DESCRIPTION
**Pull request by tchaikov**
https://trello.com/c/8f9y0YM6/51-osd-stateful-health-warnings-to-mgr-mon-eg-slow-requests

### CODE REVIEW

File: src/include/rados/librados.hpp [hunk 1]
```diff
@@ -1318,7 +1318,7 @@ namespace librados
       const std::string& name,     ///< daemon name (e.g., 'gwfoo')
       const std::map<std::string,std::string>& metadata); ///< static metadata about daemon
     int service_daemon_update_status(
-      const std::map<std::string,std::string>& status);
+      std::map<std::string,std::string>&& status);
```
Comment (tchaikov) [inline src/include/rados/librados.hpp]: @jdurgin @liewegas i understand this is part of the librados public API and is included by released tags. but my *guess* is that it is not used anywhere other than rgw, probably we can just change it w/o introducing a `service_daemon_update_status2()`?

Comment (dillaman) [reply src/include/rados/librados.hpp]: It's used by rbd-mirror and tcmu-runner (out-of-tree)

Comment (dillaman) [reply src/include/rados/librados.hpp]: ... cancel that -- tcmu-runner only utilizes the C API version.

Comment (tchaikov) [reply src/include/rados/librados.hpp]: oh, right! thanks, @dillaman:

- source code compatibility side: it's passing a temporary object. so it's passing rvalue reference already.
- binary side: we have "librados2 (= ${binary:Version})," in debian/control, and "librados2 = %{_epoch_prefix}%{version}-%{release}" in ceph.spec.

so we are safe here.

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4987,21 +4987,23 @@ void OSD::tick_without_osd_lock()
...
-      clog->warn() << *i;
+  map<string,string> daemon_status;
+  {
+    vector<string> warnings;
+    if (op_tracker.check_ops_in_flight(warnings)) {
+      daemon_status.emplace("ops_in_flight", str_join(warnings, "\n"));
     }
   }
+  if (!daemon_status.empty()) {
+    mgrc.service_daemon_update_status(std::move(daemon_status));
```
Comment (liewegas) [inline src/osd/OSD.cc]: I think the intention here was to make this map a freeform bit of state to share with the rest of the system. For health warnings, I think we want an explicit interface for passing a health_check_map_t to the mgr that gets rolled up into the map sent from mgr->mon?

**Comment by liewegas** [comment]
I guess the direction this is going, the mgr would have some special logic to translate the OSD status into health alerts.  It seems like it would be cleaner to let the OSDs do that themselves, though.. that way the logic lives in one place instead of two, and (core) daemons have a generic way of generating system-wide health alerts?

**Comment by jcsp** [comment]
The service_status piece (freeform) was meant for non-core bits, expecting that they would have their own mgr module code that understood it (i.e. the mgr is just passing it through).

For the OSD bits I think we should have a more typed mechanism, perhaps a generalisation of the MDSHealthMetric stuff in MMDSBeacon.h?

**Comment by tchaikov** [comment]
closing, i am reworking on a "more typed mechanism"

**Pull request closed without merging**